Add BLEU metric

epwalsh opened the pull request:
I thought I should share my implementation of BLEU. This is the first time I've implemented it, so there may be a more efficient way, but I'm pretty sure the math checks out at least. If anyone has thoughts on how to improve it please let me know!

42B commented on allennlp/tests/training/metrics/bleu_test.py:
```diff
@@ -0,0 +1,110 @@
+        # Unigrams.
+        counts = Counter(self.metric._ngrams(tensor, 1, set((0,))))
+        unigram_check = {(1,): 2, (2,): 2, (3,): 1}
+        assert counts == unigram_check
```
I'm curious about the use of bare assert statements inside of essentially a `unittest.TestCase` class?

```python
import unittest


class Test(unittest.TestCase):
    def test_bare(self):
        a = [1, 2, 3]
        b = [3, 4, 5]
        assert a == b

    def test_self_assert(self):
        a = [1, 2, 3]
        b = [3, 4, 5]
        self.assertEqual(a, b)


if __name__ == '__main__':
    unittest.main()

```

The output is much more descriptive by using (in this particular case) `self.assertEqual`?

```
FF
======================================================================
FAIL: test_bare (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "", line 8, in test_bare
    assert a == b
AssertionError

======================================================================
FAIL: test_self_assert (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "", line 13, in test_self_assert
    self.assertEqual(a, b)
AssertionError: Lists differ: [1, 2, 3] != [3, 4, 5]

First differing element 0:
1
3

- [1, 2, 3]
+ [3, 4, 5]

----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (failures=2)
```

epwalsh replied:
Given that we're using pytest, I think the output from a raw assert failure is actually just as, if not more, descriptive:

Here's an `assertDictEqual` failure:
![assertdictequal](https://user-images.githubusercontent.com/8812459/47880167-bfcb0900-dddf-11e8-8b9b-a67807f2cffc.png)
Here's a bare assert failure:
![base_assert](https://user-images.githubusercontent.com/8812459/47880139-a75aee80-dddf-11e8-95eb-9d9aab12db4b.png)

42B replied:
@epwalsh 

I'm curious why you would still write tests using the old unittest style then? What's the advantage?

brendan-ai2 replied:
I agree with you, @epwalsh. Bare assert seems more descriptive. It's also great to just compose raw Python instead of use custom functions.

matt-gardner commented:
@42B, we use pytest.  See https://docs.pytest.org/en/latest/.

@epwalsh, thanks, it'd be awesome to have BLEU in here!  We're mostly all at EMNLP this week, though, so it'll be a little bit before we can pick this up.

epwalsh commented:
@matt-gardner no worries! I enjoyed the slides for the writing code for NLP talk by the way 👏

maxdel commented:
Hi @epwalsh! 

So far I was wrapping [BLEU from nltk](https://www.nltk.org/api/nltk.translate.html#nltk.translate.bleu_score.sentence_bleu), since the package is anyway in allennlp's requirements.

Is there any reason about not using that one?

brendan-ai2 commented on allennlp/training/metrics/bleu.py:
```diff
@@ -0,0 +1,138 @@
+                ngram_counts[ngram] += 1
+        return ngram_counts
+
+    def _get_modified_precision(self,
```
Please add a brief comment here.

brendan-ai2 commented on allennlp/training/metrics/bleu.py:
```diff
@@ -0,0 +1,138 @@
+                if tensor_slice.size(-1) < ngram_size:
+                    break
+                ngram = tuple(x.item() for x in tensor_slice)
+                if exclude_indices and any(x in ngram for x in exclude_indices):
```
This shouldn't really matter given the size of the ngrams and `exclude_indices`, but could you swap `exclude_indices` and `ngram` so that `exclude_indices` has a lookup performed against it instead of being iterated over? Slightly more intuitive that way.

epwalsh commented:
@brendan-ai2 thanks for the review! I'll try and address your comments soon. And yes, I think you're right that the sentence brevity penalty would be wrong if we applied the NLTK implementation by batch.

epwalsh commented:
I think I addressed all your comments, @brendan-ai2, but let me know if you see anything else! Also, do you think it would be reasonable to make this the default metric for `simple_seq2seq`? That way we could get rid of this "TODO" comment: https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/simple_seq2seq.py#L213

brendan-ai2 commented:
Thanks for all the updates! Looks good overall, but mypy needs some help: http://build.allennlp.org/viewLog.html?buildId=7541&buildTypeId=AllenNLP_AllenNLPPullRequests&tab=buildLog&_focus=3050

LGTM after that!

I'd be supportive of adding BLEU as a metric for `simple_seq2seq` too. We'd probably still want to calculate the cross entropy during validation, however -- for comparison against training and to use as a stopping condition. Preferably make it configurable so that a user can disable if they don't want to pay the overhead.

epwalsh commented:
Oops, just fixed! And I agree with you on `simple_seq2seq`. I could submit a PR for that after this gets merged. Thanks!

brendan-ai2 commented:
And merged. Thanks again for the contribution! If you do modify `simple_seq2seq`, feel free to assign me as reviewer.
